keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -426,6 +431,93 @@ translation proxy terminates the WebSocket connection, and it de-multiplexes the
 various streams in order to pass the data on to a SPDY connection, which continues
 upstream (to Kubelet and eventually the container runtime).
 
+### Background: `PortForward` Subprotocol
+
+The `PortForward` subprotocol is used to implement `kubectl port-forward`, and it differs
+from the `RemoteCommand` subprotocol in how the multiple streams within the single upgraded
+connection are created. The `RemoteCommand` subprotocol statically creates streams (e.g
+STDIN, STDOUT, etc.) when the connection is created. But the `PortForward` subprotocol
+dynamically creates two streams (a bi-directional data stream and a unidirectional error
+stream) for each subsequent portforward request. These streams are removed when the request
+is complete. The connection, however, continues to exist until it is manually stopped
+(with a signal). For example, the following portforward command creates the upgraded
+connection, but without any streams, listening on the local port 8080:
+```
+$ kubectl port-forward nginx 8080:80
+```
+Once the upgraded, streaming connection is created, portforward requests are handled
+by dynamically creating a data and error stream in another goroutine, forwarding the
+data to the remote port on the target. In this example, the HTTP request is forwarded
+over the data stream from the local port 8080 to the nginx container listening on port
+80:
+```
+$ curl http://localhost:8080/index.html
+```
+The nginx HTTP response is returned over the same data stream. Once the request is
+complete, both streams are closed and removed. There is no stated limit on the number
+of requests the connection can handle, including the number of concurrent portforward
+requests. Importantly, any update to the current SPDY implementation of `PortForward`
+must also handle concurrent portforward requests.
+
+### Proposal: New `PortForward` Subprotocol Version - `v2.portforward.k8s.io`
+
+In order to implement `PortForward` over WebSockets, we propose the client attempt to
+upgrade the connection to WebSockets with a new `v2.portforward.k8s.io` subprotocol
+version. This new version will implement stream functionality over WebSockets
+which currently exists in SPDY but not WebSockets. Specifically, the WebSockets streams
+will implement:
+
+1. StreamCreate signal: A signal to the other WebSockets endpoint that a
+stream has been created. This communication will also contain the headers used to
+create the stream.
+2. StreamClose signal: A signal to the other WebSockets endpoint that the stream
+has been half-closed, and will not allow any further writing on the stream from
+the closed end.
+3. Larger stream identifier space: In order to accommodate numerous concurrent
+streams from many posssible portforward requests, we will need a large stream
+identifier space. It appears that SPDY implements 2^31 possible stream identifiers
```

Comment:
There will be one websocket connection between client and Websocket -> SPDY translator. And translator will translate `wsStreamHeader`'s to SPDY streams. I think, it is better than creating a connection per each stream requirement.

Reply:
You are correct--it should be significantly more scalable to implement many streams over a single connection, than trying to create multiple upgraded connections. Just for reference, the current SPDY implementation uses many SPDY streams over a single upgraded SPDY connection for each `kubectl port-forward...` command.